DOCS-8778 Amazon SQS 5.8 Edits 2 (#998)
mulesoft/docs-connectors: amazon-sqs/5.8/modules/ROOT/pages/amazon-sqs-connector-example-large-message.adoc

= Send and Receive Large Messages Example

This example shows two flows:

* The first flow sends a large message to the Amazon queue that references the AWS S3 bucket to use for storing large message payloads.
* The second flow receives the message and logs the message body. 

Enabling this feature incurs additional charges associated with using AWS S3.

This example uses variables for some field values. You can either replace the variables with their values in the code, or you can provide the values for each variable in the `src/main/resources/mule-artifact.properties` file.

> 

Reviewer: There are a lot of subtasks following; should we summarize them in this intro, so that the audience expects them?

== Prerequisites 

You must have:

* Access to the Amazon SQS target resource and Anypoint Platform

> * A file containing a payload larger than 256KB

Reviewer: ```suggestion
* A file containing a payload larger than 256 KB
```

[[create-mule-project]]
== Create a Mule Project

In Studio, create a new Mule project in which to add and configure the connector: 

. In Studio, select *File > New > Mule Project*.
. Enter a name for your Mule project and click *Finish*.

== Add and Configure a Listener

. In the *Mule Palette* view, search for *HTTP* and select the *Listener* operation:
+
image:amazon-sqs-select-listener.png[select-listener]
+
. Drag the *Listener* operation onto the canvas.
. In the *Listener* configuration, click *+* next to the *Connector configuration* field to add a global element.
. Accept the defaults.
. Set the *Path* field to `/largePayload`:
+
image::amazon-sqs-http-props.png[http-properties]

== Add and Configure a File Read Component 

Add and configure a *File Read* component to read the large payload:

. In the *Mule Palette* view, search for *File* and select the *File Read* operation:
+
image::amazon-sqs-file-read.png[select-file-read]
+
. Drag the *File Read* component onto the canvas, to the right of the *Listener* component.
. In the *File Read* configuration, click *+* next to the *File Configuration* field to add a global element.

> . In the File Config window, check *Connection*. 

Reviewer: ```suggestion
. In the *File Config* window, select *Connection*. 
```
check box, right?

. In *Working Directory*, enter the directory that will serve as the root directory for every relative path used with this connector. +
The default is your user home directory.
+
In this example, the file is read from `src/main/resources` of the Mule application.
+

> For example:

Reviewer: Try to avoid using this fragment GLOBALLY.

> For example:

Reviewer: ```suggestion
In this example, the file is read from `src/main/resources` of the Mule application:
```

+
image::amazon-sqs-file-global-config.png[file-global-config]
+
. Configure the following fields in the *General* properties window:
+
[%header%autowidth.spread]

> |===

Reviewer: I would like to see us move away from using tables for <thing> + <description>, toward definition lists or plain bullet lists instead (per style ref, too).

|Field |Value
|Display Name |Name for the connector operation
|Connector Configuration |Global File configuration you previously created
|File Path | Path to the file to read
|===
+

> For example:

Reviewer: Reword to avoid this fragment; GLOBAL

+
image::amazon-sqs-file-config.png[file-config]

== Add and Configure a Transform Message Component 

> Add the *Transform Message* Component to which to attach the metadata:

Reviewer: ```suggestion
Add the *Transform Message* component to which to attach the metadata:
```

----
{
	delaySeconds: 0,
	body: payload,
	messageAttributes: {
		"AccountId": {
			"stringValue" : "000123456",
			"dataType" : "String.AccountId"
		} as Object {
			class: "org.mule.extension.sqs.api.model.MessageAttributeValue"
		},
		"NumberId": {
			"stringValue" : "230.000000000000000001",
			"dataType" : "Number"
		} as Object {
			class : "org.mule.extension.sqs.api.model.MessageAttributeValue"
		}
	} as Object {
		class: "java.util.HashMap"
	}
} as Object {
	class: "org.mule.extension.sqs.api.model.Message"
}
----
+
The following screenshot shows the XML as it appears in the *Output* section of Studio:
+
image::amazon-sqs-transform-message.png[transform-message]

== Add and Configure the SQS Send Message Operation

. In the *Mule Palette* view, search for *Amazon SQS* and select the *Send message* operation:
+
image::amazon-sqs-select-send.png[select-send-message]
+
. Drag the *Send message* operation onto the canvas, to the right of the *Transform Message* component.
. In the *Send message* configuration, click *+* next to the *Connector configuration* field to add a global element.
. Configure the global element for *Send message*:
+
[%header%autowidth.spread]

> |===

Reviewer: ditto re. table
some day, if not now

|Field |Value
|Name | Name used to reference the configuration. This example uses `Amazon_SQS_Large_Payload_Configuration`.
|Session Token | Session token used to validate the temporary security credentials
|Access Key |Alphanumeric text string that uniquely identifies the user who owns the account
|Secret Key |Key that acts as a password
|Region Endpoint | Queue region
|Default Global Queue URL | Default Amazon SQS queue URL
 credentials
|===
+

> For example:

Reviewer: ditto

+
image::amazon-sqs-studio-global-config-new.png[send-global-config]
+

> . In the *Connection* section, click on the *Advanced* tab.

Reviewer: ```suggestion
. In the *Connection* section, click the *Advanced* tab.
```

. In the *Large Payload Support* field, choose either:

* *Expression or Bean Reference* 
* *Edit Inline* 
.. Configure the parameters as follows:
+
[%header%autowidth.spread]

> |===

Reviewer: ditto

|Field |Description
|Bucket | Name of the AWS S3 bucket to use for storing large message payloads. The bucket must already be created and configured in AWS S3. Enabling this feature incurs additional charges for using AWS S3.

> |Message Size Threshold | The message size threshold value for storing message payloads in the AWS S3 bucket. The default value for message size threshold is 256 KB and the maximum threshold size value is 256KB. The maximum message size is 2GB.

Reviewer: ```suggestion
|Message Size Threshold | The message size threshold for storing message payloads in the AWS S3 bucket. The default value for message size threshold is 256 KB and the maximum threshold size value is 256 KB. The maximum message size is 2 GB.
```

|Message Size Threshold Unit | Data unit for the message size threshold
|===
+

> For example:

Reviewer: ditto

+
image::amazon-sqs-studio-large-payload-support.png[Configure Large Payload Support]
+
. At the base of the Studio canvas, click *Configuration XML* to view the corresponding XML:
+
[source,dataweave,linenums]
----
<sqs:config name="Amazon_SQS_Configuration"
  doc:name="Amazon SQS Configuration"
	defaultQueueUrl="${sqs.queueUrl}">
	<sqs:basic-connection accessKey="${sqs.accessKey}"
	secretKey="${sqs.secretKey}"
	region="us-east-1"
	sessionToken="${sqs.sessionToken}">
	    <sqs:large-payload-support bucket="large-sqs-payload-bucket"
            messageSizeThreshold="256" messageSizeThresholdUnit="KB" />
    </sqs:basic-connection>
</sqs:config>
----
+
. Configure the following fields in the properties window:
+
[%header%autowidth.spread]

> |===

Reviewer: ditto

|Field |Value
|Display Name |Name for the connector operation
|Connector Configuration |Global configuration you previously created for the *Send message* operation
|Message |`payload`

> |Queue url |Amazon SQS queue URL.
> 
> If provided, the value of this field takes precedence over the value of the *Default Global Queue URL* field on the *Global Configuration Elements* window.

Reviewer: ```suggestion
|Queue url |Amazon SQS queue URL. If provided, the value of this field takes precedence over the value of the *Default Global Queue URL* field on the *Global Configuration Elements* window.
```

|===
+

> For example:

Reviewer: ditto

+
image::amazon-sqs-send-message.png[send-message]

=== Add and Configure a Logger Component

> Add and configure a *Logger* component to display the response in the Mule console:

Reviewer: The response to...? Adding and Configuring the SQS Send Message Operation, the parent task?

. In the *Mule Palette* view, search for *Logger*.
. Drag the *Logger* component onto the canvas, to the right of the *Send Message* component.
. Configure the following fields:
+
[%header%autowidth.spread]
|===
|Field |Value
|Display Name |Name for the logger, such as `Log Response`
|Message |`+++Sent Message: `#[payload]`+++`
|Level |INFO (Default)
|===
+
For example:
+
image::amazon-sqs-logger.png[logger]

=== Obtain the Number of Messages in the Queue

> 

Reviewer: Why does that matter?

. In the *Mule Palette* view, search for *Amazon SQS*.
. Select the *Get approximate number of messages* operation and drag it onto the canvas, to the right of the *Logger* component.
+
. Configure the *Queue url* field in the properties window, for example:
+
image::amazon-sqs-get-message-count.png[get-message-count]

> === Add and Configure a Logger Component

Reviewer: Add and configure a logger again?

Reviewer: ...and to achieve the same thing as the subtask just above?

Add and configure a *Logger* component to display the number of messages in the Mule console:

. In the *Mule Palette* view, search for *Logger*.
. Drag the *Logger* component onto the canvas, to the right of the *Get approximate number of messages* component.
. Configure the following fields:
+
[%header%autowidth.spread]
|===
|Field |Value
|Display Name |Name for the logger, such as `Log Count`
|Message |`+++Approx. messages in queue: `#[payload]`+++`
|Level |INFO (Default)
|===
+
For example:
+
image::amazon-sqs-logger2.png[log-message-count]

== Create a Flow to Receive Messages

Finish this example by creating another flow to receive messages and log them before they are deleted from the queue.

. In the Mule Palette view, search for *SQS* and select the *Receive messages* operation:
+
image::amazon-sqs-select-receive.png[select-receive-messages]
+
. Drag the *Receive messages* operation onto the canvas.
. Configure the following fields in the *General* properties window:
+
[%header%autowidth.spread]
|===
|Field |Value
|Display Name |Name for the connector operation

> |Connector Configuration |Global configuration you created previously `Amazon_SQS_Configuration`. If you want to receive large messages use `Amazon_SQS_Large_Payload_Configuration`.

Reviewer: ```suggestion
|Connector Configuration |Global configuration you created previously: `Amazon_SQS_Configuration`. If you want to receive large messages, use `Amazon_SQS_Large_Payload_Configuration`.
```

|Number of Messages |10

> |Queue url | Amazon SQS queue URL.
> 
> If provided, the value of this field takes precedence over the value of the *Default Global Queue URL* field on the *Global Configuration Elements* window.

Reviewer: ```suggestion
|Queue url | Amazon SQS queue URL. If provided, the value of this field takes precedence over the value of the *Default Global Queue URL* field on the *Global Configuration Elements* window.
```

|===
+
For example:
+
image::amazon-sqs-receive-message.png[receive-message]
+

> . Add a Logger component to display the message in the Mule Console.

Reviewer: ```suggestion
. Add a Logger component to display the message in the Mule console.
```
or
```suggestion
. Add a Logger component to display the message in Mule.
```
Mule Management Console is deprecated.

+
. Configure the Logger with these field values:
+
[%header%autowidth.spread]
|===
|Field |Value
|Display Name |Name of your choice
|Message |`+++`#[payload]`+++`
|Level |INFO (Default)
|===

== Example Mule Application XML Code

> 

Reviewer: By the time I got here, I had forgotten that there were two flows and one example, so I wonder if we need to reiterate something about the intro text?
```suggestion
After you complete the flow to send a large message to the Amazon queue that references the AWS S3 bucket to use for storing large message payloads, you next create a flow to receive the message and log the message body. 
```
